### PR TITLE
HPCC-19973 Add NxN test.

### PR DESCRIPTION
**Pull request by jakesmith**
created a number of paralle streams sending over a tag
with determinate data and validate on receipt.

Options to test:
-s <numstreams> (default 8)
-a <async send> (default off)
-m <MB to send> (default 100)
-b <msg size (bytes)> (default 1024)

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by jakesmith** [comment]
@mckellyln - please review

**Comment by jakesmith** [comment]
@AttilaVamos - do you know why remoteread failed? It has no relation to this PR.

**Comment by AttilaVamos** [comment]
@jakesmith Yes, I know. https://track.hpccsystems.com/browse/HPCC-20035 (PR-11380)

**Comment by jakesmith** [comment]
ok.

Re-tagging @mckellyln - please review.

File: system/mp/test/mptest.cpp [hunk 1]
```diff
@@ -777,13 +1189,17 @@ int main(int argc, char* argv[])
  *  MultiTest
  *  MPAlltoAll
  *  MPTest2
+ *  MPNxN [-b <msgSiz>] [-s <numStreams>] [-t <totalSzPerStream>] [-a]
```
Comment (mckellyln) [inline system/mp/test/mptest.cpp]: not -t but -m here

File: system/mp/test/mptest.cpp [hunk 2]
```diff
@@ -701,6 +704,411 @@ void MPAlltoAll(IGroup *group, ICommunicator *mpicomm, size32_t buffsize=0, unsi
...
+            passed = false;
+            try
+            {
+                CMessageBuffer msg, recvMsg;
+                msg.setReplyTag(replyTag);
+                void *data = msg.reserveTruncate(msgSize);
+                unsigned hash = (unsigned)mpTag;
+
+                VStringBuffer logMsg("NxN: mpTag=%u, dstRank(s) [%s]", (unsigned)mpTag, tgtRanksStr.str());
+                PROGLOG("%s", logMsg.str());
```
Comment (mckellyln) [inline system/mp/test/mptest.cpp]: PROGLOG() or PrintLog() ?  Does it matter ?  Most of the other code uses PrintLog().

File: system/mp/test/mptest.cpp [hunk 3]
```diff
@@ -701,6 +704,411 @@ void MPAlltoAll(IGroup *group, ICommunicator *mpicomm, size32_t buffsize=0, unsi
...
+                    {
+                        while (!comm->recv(msg, r, replyTag, &sender, 60000))
+                            WARNLOG("Waiting for final ack");
+                        assertex(sender == r);
+                        assertex(0 == msg.length());
+                    }
+                }
+                float ms = timer.elapsedMs();
+                float mbPerSec = (totalSendSize/ms*1000)/0x100000;
+                PROGLOG("Stream stats: time taken = %.2f seconds, total sent=%u MB, throughput = %.2f MB/s", ms/1000, (unsigned)(totalSendSize/0x100000), mbPerSec);
```
Comment (mckellyln) [inline system/mp/test/mptest.cpp]: I'm never sure, is the ms/1000 done with floating pt by default ? Or should it be (float)ms/1000.0 ?

**Review by mckellyln — APPROVED** [review]
@jakesmith Looks good, just a few minor questions

File: system/mp/test/mptest.cpp (hunk 3 above)
Comment (mckellyln) [reply system/mp/test/mptest.cpp]: ok, ms is float so I think yes

Comment (jakesmith) [reply system/mp/test/mptest.cpp]: as long as one side is float, result will be float.

File: system/mp/test/mptest.cpp (hunk 2 above)
Comment (jakesmith) [reply system/mp/test/mptest.cpp]: doesn't really matter, but PrintLog is legacy I think (but is mapped to JLOG functionality), so should be using PROGLOG or similar.

File: system/mp/test/mptest.cpp (hunk 1 above)
Comment (jakesmith) [reply system/mp/test/mptest.cpp]: will fix.

**Comment by jakesmith** [comment]
@mckellyln - please see changes.

**Comment by mckellyln** [comment]
@jakesmith looks good for squash and merge

**Comment by jakesmith** [comment]
@ghalliday - please merge.

**Pull request merged**